In [16]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

#load data
mnist = input_data.read_data_sets("MNIST_data", one_hot=True)

batch_size = 100

batch_count = mnist.train.num_examples // batch_size

#define place holder
x = tf.placeholder(tf.float32, [None, 784])
y = tf.placeholder(tf.float32, [None, 10])
keep_prob = tf.placeholder(tf.float32)

#create a 4-level neural network
W1 = tf.Variable(tf.truncated_normal([784, 500], stddev=0.1))
b1 = tf.Variable(tf.zeros([500])+0.1)
L1 = tf.nn.tanh(tf.matmul(x, W1)+b1)
L1_dropout = tf.nn.dropout(L1, keep_prob)

W2 = tf.Variable(tf.random_normal([500, 300], stddev=0.1))
b2 = tf.Variable(tf.zeros([300])+0.1)
L2 = tf.nn.tanh(tf.matmul(L1_dropout, W2)+b2)
L2_dropout = tf.nn.dropout(L2, keep_prob)

W3 = tf.Variable(tf.random_normal([300, 10], stddev=0.1))
b3 = tf.Variable(tf.zeros([10])+0.1)
prediction = tf.nn.tanh(tf.matmul(L2_dropout, W3)+b3)



#交叉熵代价函数
# loss = tf.reduce_mean(tf.square(y - prediction))
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y, logits=prediction))
# loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=prediction))

#学习率使用指数衰减法
global_step = tf.Variable(0)
learn_rate = tf.train.exponential_decay(0.001, global_step, 100, 0.95, staircase=True)

#train
# train = tf.train.GradientDescentOptimizer(0.2).minimize(loss)
train = tf.train.AdamOptimizer(learn_rate).minimize(loss, global_step=global_step)

init = tf.global_variables_initializer()

#比较预测值与实际值
correct_pred = tf.equal(tf.argmax(y, 1), tf.argmax(prediction, 1)) #argmax 返回沿着某个维度最大值的位置

#准确率
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

with tf.Session() as sess:
    sess.run(init)
    
    print("keep_prob 1.0")
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys, keep_prob:1.0})
        
        test_acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels, keep_prob:1.0})
        train_acc = sess.run(accuracy, feed_dict={x:mnist.train.images, y:mnist.train.labels, keep_prob:1.0})
        print("step:" + str(step) + " test accurancy:" + str(test_acc) + " train accurancy:" + str(train_acc) + " learning rate:" + str(sess.run(learn_rate)))



Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
keep_prob 1.0
step:0 test accurancy:0.957 train accurancy:0.96156365 learning rate:0.0007737809
step:1 test accurancy:0.9682 train accurancy:0.9762 learning rate:0.0005688001
step:2 test accurancy:0.9715 train accurancy:0.98283637 learning rate:0.00044012658
step:3 test accurancy:0.9729 train accurancy:0.9865091 learning rate:0.00032353346
step:4 test accurancy:0.9754 train accurancy:0.9876182 learning rate:0.00025034402
step:5 test accurancy:0.9771 train accurancy:0.98996365 learning rate:0.00018402585
step:6 test accurancy:0.9777 train accurancy:0.991 learning rate:0.00014239569
step:7 test accurancy:0.9778 train accurancy:0.99167275 learning rate:0.000104673905
step:8 test accurancy:0.9782 train accurancy:0.9920545 learning rate:8.0994665e-05
step:9 test accurancy:0.9785 train accurancy:0.992


keep_prob=1.0
step:0 test accurancy:0.9477 train accurancy:0.9538364
step:1 test accurancy:0.9545 train accurancy:0.9674909
step:2 test accurancy:0.9638 train accurancy:0.978
step:3 test accurancy:0.9663 train accurancy:0.9848545
step:4 test accurancy:0.9696 train accurancy:0.9872182
step:5 test accurancy:0.9718 train accurancy:0.9897636
            
keep_prob=0.7
step:0 test accurancy:0.9434 train accurancy:0.9522909
step:1 test accurancy:0.9579 train accurancy:0.96985453
step:2 test accurancy:0.9641 train accurancy:0.98014545
step:3 test accurancy:0.9677 train accurancy:0.9845273
step:4 test accurancy:0.9709 train accurancy:0.98774546
step:5 test accurancy:0.9699 train accurancy:0.9892182